File: SatellitesReactor/src/main/java/cz/radek/satellitesreactor/model/Graph.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package cz.radek.satellitesreactor.model;

import cz.radek.satellitesreactor.Main;
import cz.radek.satellitesreactor.tools.Converter;
import cz.radek.satellitesreactor.tools.GeometryTool;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.PriorityQueue;
import java.util.stream.Collectors;

/**
 *
 * @author Tomsu Radek
 */
public class Graph {

	private Node start, end;
	private final List<Node> satelliteNodes;
	private boolean initialized = false;
	private boolean calculated = false;
	private String path = "";
	private final GeometryTool geometryTool;
	private final Converter converter;
	private final double radius;

	public Graph( GeometryTool geometryTool, Converter converter, double radius ) {
		this.geometryTool = geometryTool;
		this.converter = converter;
		this.satelliteNodes = new LinkedList<>();
		this.radius = radius;
	}

	public boolean initGraph( InputData inputData ) {
		this.start = new Node( "START", converter.fromSphericalToCartesian( inputData.getStart(), radius ) );
		this.end = new Node( "END", converter.fromSphericalToCartesian( inputData.getEnd(), radius ) );
		for( Satellite satellite : inputData.getSatellites() ){
			this.satelliteNodes.add( new Node( satellite.getName(), converter.fromSphericalToCartesian( satellite.getSphericalCoordinates(), radius ) ) );
		}
		calculateEdges();
		initialized = true;
		return true;
	}

	//Use 'dijktra's' algorithm for finding shortest path as graph has positive constant valued edges
	public boolean calculateShortestPath() {
		if( !initialized ){
			return false;
		}
		PriorityQueue<Node> queue = new PriorityQueue();

		start.setDistance( 0 );
		queue.add( start );
		while( !queue.isEmpty() ){
			Node n = queue.poll();
			n.close();
			if( n.equals( end ) ){
				reconstructPath();
				calculated = true;
				return true;
			}
			for( Node neighbour : n.getNeighbours() ){
				if( neighbour.getDistance() > n.getDistance() + n.getCoordinates().distanceTo( neighbour.getCoordinates() ) ){
					queue.remove( neighbour );
					neighbour.setDistance( n.getDistance() + n.getCoordinates().distanceTo( neighbour.getCoordinates() ) );
					neighbour.setPrevNode( n );
					queue.add( neighbour );
				}
			}
		}
		return false;
	}

	public boolean isInitialized() {
		return initialized;
	}

	public boolean isCalculated() {
		return calculated;
	}

	public String getPath() {
		return path;
	}

	private void calculateEdges() {
		ArrayList<Node> temp = new ArrayList<>();
		temp.add( start );
		temp.add( end );
		temp.addAll( satelliteNodes );

		for( int i = 0; i < temp.size(); i++ ){
			for( int j = i + 1; j < temp.size(); j++ ){
				if( geometryTool.isCentreFarFromLineSegment( Cartesian.CENTRE, temp.get( i ).getCoordinates(), temp.get( j ).getCoordinates(), Main.RADIUS ) ){
					temp.get( i ).getNeighbours().add( temp.get( j ) );
					temp.get( j ).getNeighbours().add( temp.get( i ) );
				}
			}
		}
	}

	private void reconstructPath() {
		LinkedList<String> listPath = new LinkedList<>();
		Node n = end.getPrevNode();
		while( !n.equals( start ) ){
			listPath.addFirst( n.getName() );
			n = n.getPrevNode();
		}
		this.path = listPath.stream().collect( Collectors.joining( "," ) );
	}
	
	public void printCoordinates(){
		for( Node satelliteNode : satelliteNodes ){
			System.out.print( satelliteNode.getCoordinates().getX() + "; ");
		}
		System.out.println( "" );
		for( Node satelliteNode : satelliteNodes ){
			System.out.print( satelliteNode.getCoordinates().getY() + "; ");
		}
		System.out.println( "" );
		for( Node satelliteNode : satelliteNodes ){
			System.out.print( satelliteNode.getCoordinates().getZ() + "; ");
		}
	}
}

File: SatellitesCpp/src/model/Graph.cpp
/*
 * Graph.cpp
 *
 */

#include "Graph.h"

#include <math.h>
#include <functional>
#include <vector>
#include <algorithm>
#include <iostream>

#include "../tools/Converting.h"
#include "../model/InputData.h"
#include "../tools/Geometry.h"
#include "../model/Cartesian.h"

Graph::~Graph() {
	delete start_;
	delete end_;
	for (Node* satellite_node : satellite_nodes_)
		delete satellite_node;
}

bool Graph::initGraph(const InputData& inputData) {
	start_ = new Node("START",
			converting::fromSphericalToCartesian(inputData.start, radius_));
	end_ = new Node("END",
			converting::fromSphericalToCartesian(inputData.end, radius_));
	for (Satellite satellite : inputData.satellites) {
		satellite_nodes_.push_back(
				new Node(satellite.name, converting::fromSphericalToCartesian(satellite.spherical_coordinates, radius_)));
	}
	calculateEdges();
	initialized_ = true;
	return true;
}

bool Graph::calculateShortestPath() {
	if (!initialized_) {
		return false;
	}

	auto my_comp = [](Node* a, Node* b) { return a->getDistance() < b->getDistance(); };

	std::vector<Node*> node_vector;

	for(Node* node: satellite_nodes_){
		node_vector.push_back(node);
	}
	node_vector.push_back(end_);

	start_->setDistance(0);
	node_vector.push_back(start_);
	while (!node_vector.empty()) {
		Node* n = *std::min_element(node_vector.begin(), node_vector.end(), my_comp);
		node_vector.erase(std::remove(node_vector.begin(), node_vector.end(), n), node_vector.end());

		n->close();
		if (n == end_) {
			reconstructPath();
			calculated_ = true;
			return true;
		}
		for (Node* neighbour : n->getNeighbours()) {
			if (neighbour->getDistance() > n->getDistance() + n->getCoordinates().distanceTo(neighbour->getCoordinates())) {
				neighbour->setDistance(n->getDistance()	+ n->getCoordinates().distanceTo(neighbour->getCoordinates()));
				neighbour->setPrevNode(n);
			}
		}
	}
	return false;
}

void Graph::calculateEdges() {
	std::vector<Node*> temp_vector;
	temp_vector.push_back( start_ );
	temp_vector.push_back( end_ );
	for(Node* node: satellite_nodes_){
		temp_vector.push_back(node);
		}
	for( unsigned int i = 0; i < temp_vector.size(); i++ ){
		for( unsigned int j = i + 1; j < temp_vector.size(); j++ ){
			if( geometry::isCentreFarFromLineSegment( {0, 0, 0}, temp_vector[i]->getCoordinates(), temp_vector[j]->getCoordinates(), radius_ ) ){
				temp_vector[i]->getNeighbours().push_back(temp_vector[j]);
				temp_vector[j]->getNeighbours().push_back(temp_vector[i]);
			}
		}
	}
}

void Graph::reconstructPath() {
	std::list<std::string> list_path;
	Node* n = end_->getPrevNode();
	while (n != start_) {
		list_path.push_front(n->getName());
		n = n->getPrevNode();
	}
	path_ = "";
	for (auto s : list_path){
		if(!path_.empty())
			path_ += ",";
		path_ += s;
	}
}

void Graph::printCoordinates() {
	for (Node* satellite_node : satellite_nodes_) {
		std::cout << satellite_node->getCoordinates().getX() << "; " << std::endl;
	}
	std::cout << std::endl;
	for (Node* satellite_node : satellite_nodes_) {
		std::cout << satellite_node->getCoordinates().getY() << "; " << std::endl;
	}
	std::cout << std::endl;
	for (Node* satellite_node : satellite_nodes_) {
		std::cout << satellite_node->getCoordinates().getZ() << "; " << std::endl;
	}
}
